File: JS/1765_Map_Of_Highest_peak.js
var highestPeak = function(isWater) {
  const rows = isWater.length, cols = isWater[0].length

  const queue = []
  const seen = new Map()
  const result = Array(rows).fill(Infinity).map(() => Array(cols).fill(Infinity))

  convertToKey = function (...args){
    return args.join('|')
  }

  for (let row = 0; row < rows; row++){
    for (let col = 0; col < cols; col++){
      if (isWater[row][col]){
        queue.push([row, col])
        result[row][col] = 0
        seen.set(convertToKey(row, col), true)
      }
    }
  }

  const dirs = [[0, -1], [0, 1], [-1, 0], [1, 0]]
  getNei = function(row, col){
    return dirs.map(([rowStep, colStep]) => [row + rowStep, col + colStep])
                .filter(([neiRow, neiCol]) => 0 <= neiRow && neiRow < rows && 0 <= neiCol && neiCol < cols)
  }
  
  while (queue.length > 0){
    [row, col] = queue.shift()
    const neiCoords = getNei(row, col)
    neiCoords.filter(([neiRow, neiCol]) => !seen.has(convertToKey(neiRow, neiCol)))
              .forEach(([neiRow, neiCol]) => {
                seen.set(convertToKey(neiRow, neiCol), true)
                result[neiRow][neiCol] = Math.min(
                  result[neiRow][neiCol], 1 + result[row][col]
                )
                queue.push([neiRow, neiCol])
              })
  }
  return result
}

File: CPlusPlus/1765_Map_Of_Highest_peak.cpp
class Solution {
public:
    bool canMove(int nxtRow, int nxtCol, int rows, int cols) {
        return nxtRow >= 0 && nxtRow < rows && nxtCol >= 0 && nxtCol < cols;
    }
    
    vector<vector<int>> highestPeak(vector<vector<int>>& isWater) {
        int rows = isWater.size(), cols = isWater[0].size();
        vector<vector<int>> result(rows, vector<int>(cols, numeric_limits<int>::max()));
        queue<pair<int, int>> q;
        vector<vector<bool>> visited(rows, vector<bool>(cols, 0));
        vector<pair<int, int>> dirs = {
            {-1, 0},
            {0, 1},
            {1, 0},
            {0, -1}};

        for (auto row = 0; row < rows; row++) {
            for (auto col = 0; col < cols; col++) {
                if (isWater[row][col]) {
                    result[row][col] = 0;
                    visited[row][col] = 1;
                    q.push({row, col});
                }
            }
        }

        while (!q.empty()) {
            pair<int, int> topVal = q.front();
            q.pop();
            int row = topVal.first, col = topVal.second;

            for (auto& dir : dirs) {
                int nxtRow = row + dir.first, nxtCol = col + dir.second;
                if (canMove(nxtRow, nxtCol, rows, cols) && visited[nxtRow][nxtCol] == 0) {
                    visited[nxtRow][nxtCol] = 1;
                    result[nxtRow][nxtCol] = min(result[nxtRow][nxtCol], 1 + result[row][col]);
                    q.push({nxtRow, nxtCol});
                }
            }
        }
        return result;
    }
};
